scripts/o3de/o3de/repo_properties.py

```diff
@@ -0,0 +1,399 @@
+                    json_data = create_release(object_path, 
+                                   object_path / f'{object_typename}.json', 
+                                   archive_filename, 
+                                   release_archive_path, 
+                                   repo_json['repo_uri'], 
+                                   download_prefix)
+                index = _find_index(repo_objects_data, f'{object_typename}_name', json_data[f'{object_typename}_name'])
+                # if the added gem is same with partial changes
+                if index != -1:
+                    # we want changes only
+                    version_data = _changed(repo_objects_data[index], json_data)
+                    if not version_data:
+                        # there is no difference between the gem being added and 
+                        # what already exists in gems_data
+                        continue
+
+                    # versions data must always include the version
+                    version_data['version'] = json_data['version']
+
+                    # versions data must always include one of the download/source uris
+                    if 'download_source_uri' in json_data:
+                        version_data['download_source_uri'] = json_data['download_source_uri']
+                    if 'source_control_uri' in json_data:
+                        version_data['source_control_uri'] = json_data['source_control_uri']
+                    if 'source_control_ref' in json_data:
+                        version_data['source_control_ref'] = json_data['source_control_ref']
+
+                    version_index = _find_index(repo_objects_data[index].get('versions_data',[]), 'version', json_data['version'])
+                    if version_index != -1:
+                        # update the existing version
+                        if version_data:
+                            repo_objects_data[index]['versions_data'][version_index] = version_data
+                    else:
+                        # add the new version
+                        if 'versions_data' not in repo_objects_data[index]:
+                            repo_objects_data[index]['versions_data'] = [version_data] 
+                        else:
+                            repo_objects_data[index]['versions_data'].append(version_data) 
+                else:
+                    repo_objects_data.append(json_data)
+
+    # Remove duplicates from list 
+    if delete_objects:      
+        removal_list = delete_objects.split() if isinstance(delete_objects, str) else delete_objects
+        # find name field of object you want to delete
+        object_key = str(f'{object_typename}_name')
+        for removal_object in removal_list:
+            # Remove the JSON object(s) with the specified object_name value
+            repo_objects_data = [object_typename for object_typename in repo_objects_data if object_typename.get(object_key) != removal_object]
```

Comment:
we need to handle cases where the user wants to remove a specific version of an object and not every version of that object.  For example, if a repo.json has a single gem entry with 2 versions, the user should be able to just remove a single version of that gem from repo.json.  The way that command should look is something like this:
`o3de.bat repo delete-gems foo==2.0.0` 
The above command would only remove the gem with `gem_name` of `foo` and `version` `2.0.0`

If the user wanted to remove every version of a gem named `foo` they would use the command:
`o3de.bat repo delete-gems foo`
```suggestion
        removal_list = delete_objects.split() if isinstance(delete_objects, str) else delete_objects
        # find name field of object you want to delete
        object_key = str(f'{object_typename}_name')
        for removal_object in removal_list:
            object_name, version_specifier = utils.get_object_name_and_optional_version_specifier(removal_object)
            if not version_specifier :
                # Remove the JSON object(s) with the specified object_name value
                repo_objects_data = [object_typename for object_typename in repo_objects_data if object_typename.get(object_key) != object_name]
            else:
               index = _find_index(repo_objects_data, f'{object_typename}_name', object_name)
               if index > -1:
                   versions_data = repo_objects_data[index].get('versions_data',[])
                   if versions_data:
                       # remove matching versions
                       versions_data = [data for data in versions_data if Version(data.get('version', '0.0.0') not in  SpecifierSet(version_specifier)]
                       repo_objects_data[index]['versions_data'] = versions_data
                              
                   # remove/replace base version if it matches
                   if Version(repo_objects_data[index].get('version','0.0.0')) in SpecifierSet(version_specifier):
                       if versions_data:
                           # replace the base version content with all values from the first entry in versions_data
                           ...
                       else:
                           # versions data is empty so remove the whole object entry
                           del repo_objects_data[index]
               
```

Reply:
Is implementing object removal based on version a large enough task (including tests) to move to a 2nd PR? Or is it required now, because you're saying removal would be currently broken without this logic?